### ISSUE TITLE
Asynchronous mode of vLLM fails to converge.

### ISSUE THREAD

**Issue by gyr66** [volcengine/verl#2453]
Using the GSM8K dataset, I trained the Qwen-2.5-0.5B-Instruct model with GRPO. The accuracy curves on the test set with synchronous vllm and asynchronous vllm are shown below. As can be seen, the final accuracy drops when using asynchronous vllm. The version of code is v0.4.1 patch release. The asynchronous training script is as follows:

<img width="217" height="229" alt="Image" src="https://github.com/user-attachments/assets/b23701b6-2268-494c-bf3c-3a32403fa350" />

<img width="1892" height="1069" alt="Image" src="https://github.com/user-attachments/assets/e47e9766-24ea-486e-9a2e-f9e7405b3d58" />

```sh
set -x

# If you are using vllm<=0.6.3, you might need to set the following environment variable to avoid bugs:
# export VLLM_ATTENTION_BACKEND=XFORMERS

python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=$HOME/data/gsm8k/train.parquet \
    data.val_files=$HOME/data/gsm8k/test.parquet \
    data.train_batch_size=64 \
    data.max_prompt_length=512 \
    data.max_response_length=1024 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.return_raw_chat=True \
    actor_rollout_ref.model.path=Qwen/Qwen2.5-0.5B-Instruct \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=16 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.0001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.mode=async \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.8 \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger=['console','wandb'] \
    trainer.project_name='verl_grpo_example_gsm8k' \
    trainer.experiment_name='qwen2_5_05b_function_rm_async' \
    trainer.n_gpus_per_node=4 \
    trainer.nnodes=1 \
    trainer.save_freq=10 \
    trainer.test_freq=5 \
    trainer.validation_data_dir=$HOME/validation \
    trainer.default_local_dir=/home/guoyiran/data/checkpoints/verl_grpo_example_gsm8k_async \
    trainer.max_actor_ckpt_to_keep=1 \
    trainer.max_critic_ckpt_to_keep=1 \
    trainer.total_epochs=4 | tee $HOME/training.log 
```

**Comment by casper-hansen** [comment]
Hi @wuxibin89. Is this PR potentially related to this comment https://github.com/volcengine/verl/pull/1946#issuecomment-2995459715? If yes, I believe it would be good to deprecate ChatCompletion and move the async implementation to the SingleTurn agent.

**Comment by wuxibin89** [comment]
@casper-hansen Yes, please use the new AgentLoop implementation, both `SingleTurnAgentLoop` and `ToolAgentLoop` accuracy have been verified on [BytedTsinghua-SIA/DAPO-Math-17k](https://huggingface.co/datasets/BytedTsinghua-SIA/DAPO-Math-17k). The gsm8k dataset is too simple to verify the accuracy.

**Comment by gyr66** [comment]
@wuxibin89 I tested the implementation of AgentLoop from the latest main branch and found that there is still a difference in accuracy between the asynchronous and synchronous modes.

<img width="636" height="360" alt="Image" src="https://github.com/user-attachments/assets/a8411c11-fe86-4889-95a4-078fbe9a4dd2" />

**Comment by wuxibin89** [comment]
@gyr66 How about the `critic/rewards` metrics?

**Comment by gyr66** [comment]
@wuxibin89 

<img width="634" height="360" alt="Image" src="https://github.com/user-attachments/assets/85147b4a-d10a-4e12-9bbd-fc30cebc4625" />

<img width="621" height="340" alt="Image" src="https://github.com/user-attachments/assets/89982b1e-6a0c-4ea9-be65-fbff7762d77f" />

Score seems very close. The advantage of async mode is slightly higher than the sync mode.

**Comment by gyr66** [comment]
@wuxibin89 I previously encountered a segment fault error, which caused the red line to end prematurely. I’m not sure what the reason was. After recovery (the green line in the figure), I found that the peak points are quite close. I think this might be normal?

<img width="1223" height="352" alt="Image" src="https://github.com/user-attachments/assets/4756f12c-67c3-4901-bbdb-50f564d14e51" />

<img width="639" height="355" alt="Image" src="https://github.com/user-attachments/assets/378b10ae-d105-4c29-a94d-f852eb987af1" />

**Comment by wuxibin89** [comment]
@gyr66 From the green line and blue line, I think they're pretty close.

**Comment by xidongyuandong** [comment]
> [@wuxibin89](https://github.com/wuxibin89) I previously encountered a segment fault error, which caused the red line to end prematurely. I’m not sure what the reason was. After recovery (the green line in the figure), I found that the peak points are quite close. I think this might be normal?
> 
> <img alt="Image" width="1223" height="352" src="https://private-user-images.githubusercontent.com/87962456/465138526-4756f12c-67c3-4901-bbdb-50f564d14e51.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MzExMjEsIm5iZiI6MTc2NzUzMDgyMSwicGF0aCI6Ii84Nzk2MjQ1Ni80NjUxMzg1MjYtNDc1NmYxMmMtNjdjMy00OTAxLWJiZGItNTBmNTY0ZDE0ZTUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDEyNDcwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4M2Q4NTU3ZGVmMjI2ZjU2NzUyYmJiMjM5NzMxNjg5MmFjZWJmNmY2NWZjOWI3NDIxZTY4NzRkYjljN2NkZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WJCj7k35ye4CKEw3a6yORdLm4OpWexOcz2Xj81RoUBo"> <img alt="Image" width="639" height="355" src="https://private-user-images.githubusercontent.com/87962456/465138301-378b10ae-d105-4c29-a94d-f852eb987af1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MzExMjEsIm5iZiI6MTc2NzUzMDgyMSwicGF0aCI6Ii84Nzk2MjQ1Ni80NjUxMzgzMDEtMzc4YjEwYWUtZDEwNS00YzI5LWE5NGQtZjg1MmViOTg3YWYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDEyNDcwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5NTNlYTNhNzIwNTAyOTIxNzQ5ZWU5NDVkMzhhZTdiZTMwZjFlNmQ3MzEzNjVmZTBmMDAxNWJlMmNlYzU4ZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SeJ18cckCnGudMiGr6Ujahg76hyp27POnZH19ZoombU">

after segment fault,can you still train?

**Comment by Kirrito-k423** [comment]
### 中文
定时清理检查结论：这个 issue 最近一次非当前运行账号的公开回复或引用是 `2026-01-04T12:47:16Z`，截至本轮运行时间 `2026-07-10T01:03:01Z` 已经超过 6 个月。检查来源包括 issue comments API `https://api.github.com/repos/verl-project/verl/issues/2453/comments` 和 timeline API `https://api.github.com/repos/verl-project/verl/issues/2453/timeline`；最近证据：https://github.com/verl-project/verl/issues/2453#issuecomment-3708057941。

因此按社区 issue 定时清理标准关闭。若作者或社区成员仍能提供当前版本的复现、日志、PR 关联或新的需求背景，欢迎继续评论请求重开，或创建一个带有最新证据的新 issue。

### English
Scheduled cleanup check: the latest public reply or reference from an account other than the current automation account was `2026-01-04T12:47:16Z`, which is more than 6 months before this run time `2026-07-10T01:03:01Z`. The checked public sources were the issue comments API `https://api.github.com/repos/verl-project/verl/issues/2453/comments` and the timeline API `https://api.github.com/repos/verl-project/verl/issues/2453/timeline`; latest evidence: https://github.com/verl-project/verl/issues/2453#issuecomment-3708057941.

Closing this under the scheduled issue cleanup policy. If the author or community can still provide a current reproduction, logs, PR reference, or updated requirement context, please comment to request reopening or open a new issue with fresh evidence.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
